Models/IonAuthModel.php

```diff
@@ -0,0 +1,2798 @@
+
+		// Retrieve the user according to this selector
+		$user = $this->where('forgotten_password_selector', $token->selector)->users()->row();
+
+		if ($user)
+		{
+			// Check the hash against the validator
+			if ($this->verifyPassword($token->validator, $user->forgotten_password_code))
+			{
+				return $user;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Register
+	 *
+	 * @param string $identity
+	 * @param string $password
+	 * @param string $email
+	 * @param array  $additional_data
+	 * @param array  $groups
+	 *
+	 * @return integer|boolean
+	 * @author Mathew
+	 */
+	public function register(string $identity, string $password, string $email, array $additional_data = [], array $groups = [])
+	{
+		$this->triggerEvents('pre_register');
+
+		$manualActivation = $this->config->manualActivation;
+
+		if ($this->identityCheck($identity))
+		{
+			$this->setError('IonAuth.account_creation_duplicate_identity');
+			return false;
+		}
+		else if (! $this->config->defaultGroup && empty($groups))
+		{
+			$this->setError('IonAuth.account_creation_missing_defaultGroup');
+			return false;
+		}
+
+		// check if the default set in config exists in database
+		$query = $this->db->table($this->tables['groups'])->where(['name' => $this->config->defaultGroup], 1)->get()->getRow();
+		if (!isset($query->id) && empty($groups))
+		{
+			$this->setError('IonAuth.account_creation_invalid_defaultGroup');
+			return false;
+		}
+
+		// capture default group details
+		$defaultGroup = $query;
+
+		// IP Address
+		$ip_address = \Config\Services::request()->getIPAddress();
+
+		// Do not pass $identity as user is not known yet so there is no need
+		$password = $this->hashPassword($password);
+
+		if ($password === false)
+		{
+			$this->setError('IonAuth.account_creation_unsuccessful');
+			return false;
+		}
+
+		// Users table.
+		$data = [
+			$this->identity_column => $identity,
+			'username' => $identity,
+			'password' => $password,
+			'email' => $email,
+			'ip_address' => $ip_address,
+			'created_on' => time(),
+			'active' => ($manualActivation === false ? 1 : 0)
+		];
+
+		// filter out any data passed that doesnt have a matching column in the users table
+		// and merge the set user data and the additional data
+		$user_data = array_merge($this->_filterData($this->tables['users'], $additional_data), $data);
+
+		$this->triggerEvents('extra_set');
+
+		$this->db->table($this->tables['users'])->insert($user_data);
+
+		$id = $this->db->insertId($this->tables['users'] . '_id_seq');
+
+		// add in groups array if it doesn't exists and stop adding into default group if default group ids are set
+		if (isset($defaultGroup->id) && empty($groups))
+		{
+			$groups[] = $defaultGroup->id;
+		}
+
+		if (!empty($groups))
+		{
+			// add to groups
+			foreach ($groups as $group)
+			{
+				$this->addToGroup($group, $id);
+			}
+		}
+
+		$this->triggerEvents('post_register');
+
+		return (isset($id)) ? $id : false;
```

Comment:
Use null coalescing operator
```suggestion
		return $id ?? false;
```